pgmpy/models/DiscreteBayesianNetwork.py

```diff
@@ -1190,9 +1200,28 @@ def do(self, nodes, inplace=False):
         adj_model = DAG.do(model, nodes, inplace=inplace)
 
         if adj_model.cpds:
-            for node in nodes:
+            for index, node in enumerate(nodes):
                 cpd = adj_model.get_cpds(node=node)
-                cpd.marginalize(cpd.variables[1:], inplace=True)
+                state_names = cpd.state_names[node]
+                if states[index] not in state_names:
```

Comment:
We usually try to do all the checks at the beginning of functions; it can help with readability as it separates outs checks and actual code logic. So, I would suggest putting this somewhere at the start of the method with the other checks.